### DISCUSSION TITLE
Why is the strides of conv_block 1 in stage 2 of ResNet50?

### DISCUSSION THREAD

**Discussion by parkjjoe** [community/community#61512] [Programming Help]
This is a part of the ResNet50 code.

```
def ResNet50(input_shape=(224, 224, 3), classes=10):
    img_input = layers.Input(shape=input_shape)

    x = layers.ZeroPadding2D(padding=(3, 3))(img_input)
    x = layers.Conv2D(64, (7, 7), strides=(2, 2), padding='valid')(x)
    x = layers.BatchNormalization(axis=3)(x)
    x = layers.Activation('relu')(x)
    x = layers.MaxPooling2D((3, 3), strides=(2, 2))(x)

// stage 2
    x = conv_block(x, 3, [64, 64, 256], strides=(1, 1))
    x = identity_block(x, 3, [64, 64, 256])
    x = identity_block(x, 3, [64, 64, 256])

    x = conv_block(x, 3, [128, 128, 512])
    x = identity_block(x, 3, [128, 128, 512])
    x = identity_block(x, 3, [128, 128, 512])
    x = identity_block(x, 3, [128, 128, 512])

    x = conv_block(x, 3, [256, 256, 1024])
    x = identity_block(x, 3, [256, 256, 1024])
    x = identity_block(x, 3, [256, 256, 1024])
    x = identity_block(x, 3, [256, 256, 1024])
    x = identity_block(x, 3, [256, 256, 1024])
    x = identity_block(x, 3, [256, 256, 1024])

    x = conv_block(x, 3, [512, 512, 2048])
    x = identity_block(x, 3, [512, 512, 2048])
    x = identity_block(x, 3, [512, 512, 2048])

    x = layers.GlobalAveragePooling2D(name='avg_pool')(x)
    x = layers.Dense(classes, activation='softmax')(x)

    model = tf.keras.Model(img_input, x)

    return model
```

One of the differences between `conv_block` and `identity_block` is 'strides', and `conv_block` uses basically `strides=2`, so I wonder why the strides of `conv_block` is 1 in stage 2.

If the strides of `conv_block` is 1, isn't it okay to write `identity_block` instead? But why are most ResNet50 codes written in `conv_block`?

**Comment by Awrsha** [comment]
The conv_block with strides=1 in stage 2 of ResNet50 is used to increase the number of channels, while keeping the spatial dimensions the same. This allows the following identity_blocks to operate on that larger channel dimension.

Identity blocks only perform convolutions with strides=1, so they keep the spatial dimensions and channel dimension the same.

So in stage 2 of ResNet50, we have:

conv_block with strides=1: This increases the channel dimension from 256 to 512, while keeping the spatial dimensions the same.

Then 3 identity_blocks operate on this 512 channel feature map. They perform convolutions with strides=1, keeping the dimensions unchanged.

This pattern of an initial conv_block followed by several identity_blocks is repeated for each stage of ResNet. The conv_block increases the channel dimension to allow more representational power, and then the identity_blocks perform that representation on that larger channel dimension.

So in summary:

conv_block with strides=1 is used to increase the channel dimension
identity_block is used to further represent that larger channel dimension, without changing dimensions
Both block types are needed - conv_block to progressively increase channels, and identity_block to take advantage of those increased channels.
